query-engine/connectors/mongodb-query-connector/src/root_queries/update_utils.rs

```diff
@@ -0,0 +1,127 @@
+use super::*;
+use crate::IntoBson;
+use connector_interface::{DatasourceFieldName, NestedWrite, WriteExpression};
+use itertools::Itertools;
+use mongodb::bson::{doc, Document};
+use prisma_models::PrismaValue;
+
+pub fn render_update_docs(
+    write_expr: WriteExpression,
+    field: &Field,
+    field_name: &str,
+) -> crate::Result<Vec<Document>> {
+    if let WriteExpression::NestedWrite(nested_write) = write_expr {
+        let docs = unfold_nested_write(field, nested_write, &mut vec![field.db_name().to_owned()])
+            .into_iter()
+            // TODO: figure out why we can't flat_map here
+            // TODO: the trait `FromIterator<Vec<mongodb::bson::Document>>` is not implemented for `std::result::Result<Vec<mongodb::bson::Document>, MongoError>`
+            .map(|(write_expr, field, field_name)| render_update_docs(write_expr, field, &field_name))
+            .collect::<crate::Result<Vec<_>>>()?;
+
+        return Ok(docs.into_iter().flatten().collect_vec());
+    };
```

Comment:
Do you have any idea what's up here?

Reply:
Additional note: this whole file is refactored in a follow-up PR.

Reply:
Check out the signature of `flat_map`: In essence, it tells you that instead of returning single elements in `map`, flat map returns an `IntoIterator` that will be consumed to flatten. 

`render_update_docs` returns a `Result` and result types in rust are `IntoIterator`! They return 0 elements if they are `Err`, 1 if they are `OK`, so writing `.flat_map(|(write_expr, field, field_name)| render_update_docs(write_expr, field, &field_name))` is actually fine. What is NOT fine is then trying to collect into `Result<Vec<_>>`. Why? Because the result is already consumed during the flattening because internally, `flat_map` calls `into_iter` and handles the magic for you. Since you don't have `Result` anymore, you _could_ collect like this: ` .collect::<Vec<_>>();`, but note with this approach you're losing all error information as errors are simply zero-length iterators. 

Tl;dr: You can't use `flat_map` here because that would lose error info. You need a two-step approach. You could also rewrite it into a for loop and not iterate twice.

Reply:
Oh lord, that wasn't trivial. Super useful explanation, thank you 🙏. I'll refactor that to a for-loop in the follow-up PR if you don't mind 👍